---
title: "DESeq2"
author: "allyson_demerlis"
date: "2023-08-14"
output: html_document
---

```{r}
library(tidyverse)
library(DESeq2)
library(genefilter) #for filterfun and genefilter
library(factoextra) #for PCA and eigenvectors
library(vegan)
library(pheatmap)
```

Import and tidy gene counts matrix
```{r}
genecounts <- read.csv("../../results/Acer/gene_count_acerv_matrix.csv")
genecounts %>% column_to_rownames(var = "gene_id") -> genecounts

#tidy column names
colnames(genecounts) -> genecounts_colnames

gsub("^(.*?)_.*$", "\\1", genecounts_colnames) -> new_colnames

for (i in seq_along(new_colnames)) {
  colnames(genecounts)[colnames(genecounts) == colnames(genecounts)[i]] <- new_colnames[i]
}
```

import aligned reads sample metadata
```{r}
sequencing_data <- read.csv("../../RNA_extraction_sequencing_data.csv")

#obtaining list of Acer samples which passed read count
sequencing_data %>% 
  select(Species:Treatment,Sequence.File.Sample.Name,Raw.Reads..Million.,M.Reads.After.Filtering..Trimmed.reads.,Mapping.Rate..Post.trimming.alignment.) %>% 
  mutate(percent_alignment = gsub("%","", Mapping.Rate..Post.trimming.alignment.)) %>% 
  mutate(percent_alignment = as.numeric(percent_alignment)) %>% 
  mutate(percent_alignment = percent_alignment / 100) %>% 
  mutate(million_reads_aligned = (M.Reads.After.Filtering..Trimmed.reads.)*percent_alignment) %>% 
  filter(million_reads_aligned > 4) %>% 
  select(Species:Treatment,Sequence.File.Sample.Name,million_reads_aligned) %>% 
  filter(Species == "Acer") %>% 
  mutate(Sequence.File.Sample.Name = gsub("^(.*?)_.*$", "\\1", Sequence.File.Sample.Name)) %>% 
  mutate(Sequence.File.Sample.Name = gsub("-", ".", Sequence.File.Sample.Name)) -> Acer_samples_4M

#which ones need to be filtered out?
sequencing_data %>% 
  select(Species:Treatment,Sequence.File.Sample.Name,Raw.Reads..Million.,M.Reads.After.Filtering..Trimmed.reads.,Mapping.Rate..Post.trimming.alignment.) %>% 
  mutate(percent_alignment = gsub("%","", Mapping.Rate..Post.trimming.alignment.)) %>% 
  mutate(percent_alignment = as.numeric(percent_alignment)) %>% 
  mutate(percent_alignment = percent_alignment / 100) %>% 
  mutate(million_reads_aligned = (M.Reads.After.Filtering..Trimmed.reads.)*percent_alignment) %>% 
  filter(Species == "Acer") %>% 
  filter(million_reads_aligned < 4)
#Acer-072, Acer-089, Acer-108
```

Check that there are no genes with 0 counts across all samples
```{r}
dim(genecounts) #33715    48
nrow(Acer_samples_4M) #45
#3 samples were filtered out

#Acer-072, Acer-089, Acer-108 from above

#manually remove them because you couldn't figure out how to use code to do it

samples_to_remove <- c("Acer.072", "Acer.089", "Acer.108")
genecounts_filt <- genecounts %>% select(-one_of(samples_to_remove))

#filter out counts that have zero
genecounts_filt <-genecounts_filt %>%
    mutate(Total = rowSums(.[, 1:45]))%>%
    filter(!Total==0)%>%
    dplyr::select(!Total)
nrow(genecounts_filt)

#went from 33715 to 14728 genes
```

Filter reads by proportion of samples containing cutoff value
```{r}
filt <- filterfun(pOverA(0.85,5)) #this means keep 85% of genes that have a count >5

tfil <- genefilter(genecounts_filt, filt)

keep <- genecounts_filt[tfil,]

gn.keep <- rownames(keep)

genecounts_filt <- as.data.frame(genecounts_filt[which(rownames(genecounts_filt) %in% gn.keep),])

#now at 4146 genes

#write.csv(genecounts_filt, "../../results/Acer/genecounts_filtered.csv")
```

Quality check of datasets to make sure row and column names match
```{r}
Acer_samples_4M %>% 
  column_to_rownames(var = "Sequence.File.Sample.Name") -> Acer_samples_4M

all(rownames(Acer_samples_4M) %in% colnames(genecounts_filt)) #TRUE 
all(rownames(Acer_samples_4M) == colnames(genecounts_filt)) # FALSE
```

Display order of metadata and gene count matrix, and reorder columns of genecounts_filt so they match the samples (need to do this before creating DESeq2 object because the rows and columns need to match EXACTLY)
```{r}
rownames(Acer_samples_4M)
colnames(genecounts_filt)

desired_order <- as.vector(rownames(Acer_samples_4M))

genecounts_filt %>% select(matches(desired_order)) -> genecounts_filt_ordered

all(rownames(Acer_samples_4M) == colnames(genecounts_filt_ordered)) #TRUE!
```


set variables in metadata as factors
```{r}
Acer_samples_4M$Genotype <- factor(Acer_samples_4M$Genotype)
Acer_samples_4M$Treatment <- factor(Acer_samples_4M$Treatment)

#change genotype names to not have hyphens because DESeq2 doesn't like that
Acer_samples_4M$Genotype <- gsub("-", "_", Acer_samples_4M$Genotype)
#also need to get rid of spaces I think
Acer_samples_4M$Genotype <- gsub(" ", "", Acer_samples_4M$Genotype)

#make a new column for "Day_0" or "Day_29" to correspond to beginning and end sampling time points
Acer_samples_4M %>% 
  mutate(time_point = case_when(Experiment.phase == "Pre-treatment" ~ "Day_0",
                                Experiment.phase == "last day of treatment" ~ "Day_29")) -> Acer_samples_4M

Acer_samples_4M$time_point <- factor(Acer_samples_4M$time_point)
```

## DESeq accounting for all variables

create matrix for DESeq
```{r}
data <- DESeqDataSetFromMatrix(countData = genecounts_filt_ordered, colData = Acer_samples_4M, design = ~ Treatment*time_point + Genotype)
```

estimate size factors
```{r}
SF.data <- estimateSizeFactors(data)
print(sizeFactors(SF.data)) # everything is less than 4, so I can use vst
```


Apply variance stabilizing transformation to minimize effects of small counts and normalize wrt library size
```{r}
vst <- vst(data, blind = FALSE) #accounts for within group variability
```

scree plot for variance
```{r}
pca <- prcomp(t(assay(vst)))
fviz_eig(pca)
```

PCA
```{r}
plotPCA(vst, intgroup = c("Treatment"))
plotPCA(vst, intgroup = c("Genotype"))
plotPCA(vst, intgroup = c("time_point"))

vst_PCAdata <- plotPCA(vst, intgroup = c("Treatment", "Genotype", "time_point"), returnData = TRUE)
percentVar <- round(100*attr(vst_PCAdata, "percentVar")) 

#plot PCA of samples with all data
pca.centroids <- vst_PCAdata %>% 
  dplyr::select(Treatment, Genotype, time_point, PC1, PC2)%>%
  dplyr::group_by(Treatment, Genotype, time_point)%>%
  dplyr::summarise(PC1.mean = mean(PC1),
                   PC2.mean = mean(PC2))
find_hull <- function(vst_PCAdata) vst_PCAdata[chull(vst_PCAdata$PC1, vst_PCAdata$PC2), ]
hulls <- plyr::ddply(vst_PCAdata, "group", find_hull)


ggplot(vst_PCAdata, aes(PC1, PC2, color=Genotype, shape=time_point)) + 
   geom_point(size=3) +
   xlab(paste0("PC1: ",percentVar[1],"% variance")) +
   ylab(paste0("PC2: ",percentVar[2],"% variance")) +
   ggtitle("A. cervicornis - all genes") +
   theme_bw() + #Set background color
   theme(panel.border = element_blank(), # Set border
         axis.line = element_line(colour = "black"), #Set axes color
         plot.background=element_blank()) + 
  stat_ellipse(aes(PC1, PC2, group=Treatment, lty = Treatment), type = "norm")

#trying to group things by time_point x genotype x treatment combo
ggplot(vst_PCAdata, aes(PC1, PC2, color=Treatment, shape=Treatment)) + 
   geom_point(size=3) +
   xlab(paste0("PC1: ",percentVar[1],"% variance")) +
   ylab(paste0("PC2: ",percentVar[2],"% variance")) +
   ggtitle("A. cervicornis - all genes") +
   theme_bw() + #Set background color
   theme(panel.border = element_blank(), # Set border
         axis.line = element_line(colour = "black"), #Set axes color
         plot.background=element_blank()) + #Set the plot background
  geom_polygon(data = hulls, alpha = 0.2, aes(color = group, fill = group))
```


Run DESeq2 
```{r}
data <- DESeqDataSetFromMatrix(countData = genecounts_filt_ordered, colData = Acer_samples_4M, design = ~ Treatment*time_point + Genotype)
```

Conduct PERMANOVA
```{r}
test<-t(assay(vst))
test<-as.data.frame(test)

test$Sample_ID <-rownames(test)
test$Treatment <- Acer_samples_4M$Treatment[match(test$Sample_ID, rownames(Acer_samples_4M))]
test$Genotype <- Acer_samples_4M$Genotype[match(test$Sample_ID, rownames(Acer_samples_4M))]
test$time_point <- Acer_samples_4M$time_point[match(test$Sample_ID, rownames(Acer_samples_4M))]
```

Build PERMANOVA model
```{r}
dim(test) #45 4150
scaled_test <-prcomp(test[c(1:4146)], scale=TRUE, center=TRUE) #subtract 4 from the dim because you made 4 columns for metadata
fviz_eig(scaled_test)
vegan <- scale(test[c(1:4146)])

permanova<-adonis2(vegan ~ Treatment*time_point + Genotype, data = test, method = 'eu')

print(permanova) #everything is significant woooooo
```

Run DESeq2
```{r}
DEG_all <- DESeq(data)
DEG_all_res <- results(DEG_all, alpha = 0.05)
summary(DEG_all_res)

resultsNames(DEG_all)

DEG_treatment <- results(DEG_all, contrast = c("Treatment", "variable", "control"), alpha = 0.05)
summary(DEG_treatment)

DEG_timepoint <- results(DEG_all, contrast = c("time_point", "Day_29", "Day_0"), alpha = 0.05)
summary(DEG_timepoint)

# i don't think these results are informative. I want a result that says "variable versus control at day 29 (so after the treatment to get an effect of treatment?)". or do I want to also see what they looked like at day 0? or do I do an LRT where I use day 0 as the "baseline" and then compare the effect of treatment after the 29 days. I don't necessarily care about the differential gene expression on day 0. That would be the sort of "baseline" gene expression differences between genotypes 
```

DESeq2 with LRT
```{r}
data <- DESeqDataSetFromMatrix(countData = genecounts_filt_ordered, colData = Acer_samples_4M, design = ~ Treatment*time_point + Genotype)

ddsLRT <- DESeq(data, test="LRT", reduced = ~Treatment + time_point)

 #this reduced function is including treatment + time_point (but not the interaction), which means than any differences at time point 0 will be controlled for. (see Michael Love's response on this thread: https://support.bioconductor.org/p/62684/  "Using a reduced formula of ~ time + treat means that genes which show a consistent difference from time 0 onward will not have a small p value. This is what we think you want, because differences between groups at time 0 should be controlled for, e.g. random differences between the individuals chosen for each treatment group which are observable before the treatment takes effect.")

resLRT <- results(ddsLRT)
summary(resLRT, alpha = 0.05)

resultsNames(ddsLRT)

#identifying significant genes
resSig_LRT<-resLRT[which(resLRT$padj<0.05), ]
```



## DESeq separating time points - just genotype x treatment in each formula

# Day 0

filter for just the first time point
```{r}
Acer_samples_4M %>% filter(time_point == "Day_0") -> day_0_samples

list_day0 <- rownames(day_0_samples)

genecounts_filt_ordered %>% select(matches(list_day0)) -> day_0_genecounts
```

create matrix for DESeq
```{r}
day0_dds <- DESeqDataSetFromMatrix(countData = day_0_genecounts, colData = day_0_samples, design = ~ Treatment + Genotype)
```

estimate size factors
```{r}
SF_day0 <- estimateSizeFactors(day0_dds)
print(sizeFactors(SF_day0)) # everything is less than 4, so I can use vst
```


Apply variance stabilizing transformation to minimize effects of small counts and normalize wrt library size
```{r}
vst_day0 <- vst(day0_dds, blind = FALSE) #accounts for within group variability
```

scree plot for variance
```{r}
pca_day0 <- prcomp(t(assay(vst_day0)))
fviz_eig(pca_day0)
```

PCA
```{r}
plotPCA(vst_day0, intgroup = c("Treatment"))
plotPCA(vst_day0, intgroup = c("Genotype"))

vst_day0_PCAdata <- plotPCA(vst_day0, intgroup = c("Treatment", "Genotype"), returnData = TRUE)
percentVar <- round(100*attr(vst_day0_PCAdata, "percentVar")) 

#plot PCA of samples with all data
pca.centroids_day0 <- vst_day0_PCAdata %>% 
  dplyr::select(Treatment, Genotype, PC1, PC2)%>%
  dplyr::group_by(Treatment, Genotype)%>%
  dplyr::summarise(PC1.mean = mean(PC1),
                   PC2.mean = mean(PC2))
find_hull_day0 <- function(vst_day0_PCAdata) vst_day0_PCAdata[chull(vst_day0_PCAdata$PC1, vst_day0_PCAdata$PC2), ]
hulls_day0 <- plyr::ddply(vst_day0_PCAdata, "group", find_hull_day0)

# first PCA
ggplot(vst_day0_PCAdata, aes(PC1, PC2, color=Genotype, shape=Treatment)) + 
   geom_point(size=3) +
   xlab(paste0("PC1: ",percentVar[1],"% variance")) +
   ylab(paste0("PC2: ",percentVar[2],"% variance")) +
   ggtitle("Day 0 Acer - all genes") +
   theme_bw() + #Set background color
   theme(panel.border = element_blank(), # Set border
         axis.line = element_line(colour = "black"), #Set axes color
         plot.background=element_blank())
```


```{r}
# attempting to create polygons around treatment but it looks weird
ggplot(vst_day0_PCAdata, aes(PC1, PC2, color=Treatment)) + 
   geom_point(size=3, aes(shape=Genotype)) +
  scale_color_manual(labels=c("Control", "Variable"), values = c( "#60DBDB", "#F54A34")) +
   xlab(paste0("PC1: ",percentVar[1],"% variance")) +
   ylab(paste0("PC2: ",percentVar[2],"% variance")) +
   ggtitle("Day 0 Acer - all genes") +
   theme_bw() + #Set background color
geom_polygon(data = hulls_day0, alpha = 0.2, aes(color = Treatment, fill = Treatment)) + #Add centroids
  geom_point(aes(x=PC1.mean, y=PC2.mean,color=Treatment, shape = Genotype), data=pca.centroids_day0, size=4, show.legend=FALSE)
```


```{r}
# attempting to create polygons around genotype but that also looks weird
ggplot(vst_day0_PCAdata, aes(PC1, PC2, color=Genotype)) + 
   geom_point(size=3, aes(shape=Treatment)) +
   xlab(paste0("PC1: ",percentVar[1],"% variance")) +
   ylab(paste0("PC2: ",percentVar[2],"% variance")) +
   ggtitle("Day 0 Acer - all genes") +
   theme_bw() + #Set background color
geom_polygon(data = hulls_day0, alpha = 0.2, aes(color = Genotype, fill = Genotype)) + #Add centroids
  geom_point(aes(x=PC1.mean, y=PC2.mean,color=Genotype, shape = Treatment), data=pca.centroids_day0, size=4, show.legend=FALSE) 
```

# Day 29
