## Disclaimer:

There are three completely different pipelines that exist within this "gene_expression" directory, which came from trials and tribulations of trying to understand different pipelines and programs.

My most up-to-date pipeline that I plan to publish with is the "MS_Bioinformatics" folder, which originated from [Dr. Michael Studivan](https://github.com/mstudiva). 

Within the "archive" folder, there are two other pipelines. The "HP_pipeline" comes from Dr. Hollie Putnam's lab, and the "NJ_Pipeline" came from [Dr. Natalia Andrade](https://github.com/China2302) and [Jill Ashey](https://github.com/JillAshey). Ultimately I tried many different programs, but Michael helped me identify PCR duplicates with his custom perl/python scripts which were adapted from [Dr. Mikhail Matz](https://github.com/z0on), who developed 3' TagSeq for corals (I think [Meyer et al. 2011](https://onlinelibrary.wiley.com/doi/10.1111/j.1365-294X.2011.05205.x) is the publication for that) and we had gotten our samples sequenced from the same facility that he works at, [UT Austin GSAF](https://wikis.utexas.edu/display/GSAF/About+Us). All this is to say, there is no one "correct" pipeline, just an abundance of different tools that seem to work best depending on the species used, type of sequencing performed, quality of sequencing data, and quality/availability of your species' genome or transcriptome. 

I wanted to credit everyone who has helped me along the way as I learn bioinformatics and gene expression analysis, and provide links to GitHubs that had well-annotated code that helped me immensely!
