---
title: "specific_gene_expression"
author: "allyson_demerlis"
date: "2024-03-06"
output: html_document
---

# load libraries and data
```{r}
library(tidyverse)
library(DESeq2)
library(cowplot)

TreatedvUntreated <- read_csv("results_csv/unique_Treated_vs_Untreated_annotated_KOG.csv")

load("Rdata_files/initial_fullddsdesigncountsVsdcountsWGCNA.RData")
load("Rdata_files/vsd.RData")
```

# vst normalized counts
```{r}
metadata <- read_csv("../../physiology/metadata.csv")

metadata %>% 
  dplyr::select(Species, ID, Treatment) %>% 
  filter(Species == "Acropora cervicornis") %>% 
  dplyr::select(!Species) %>% 
  dplyr::rename(Sample_ID = ID) -> metadata

metadata$Formatted <- sprintf("Acer.%03d", metadata$Sample_ID)

metadata %>% 
  drop_na() %>%  
  dplyr::select(!Sample_ID) %>% 
  dplyr::rename(Sample_ID = Formatted) -> metadata


TreatedvUntreated %>% 
  arrange(desc(log2FoldChange)) #Acropora_023112 = histone H2A l2fc 1.48 

TreatedvUntreated %>% 
  arrange(log2FoldChange) #Acropora_019471 = 	nitric-oxide synthase activity l2fc -2.052

standard_error <- function(x) {
  sd(x) / sqrt(length(x))
}
```

## histone 2A linear plot
```{r}
#Acropora_023112 = histone H2A l2fc 1.48 
plot1<-t(vsd) %>% 
  as.data.frame() %>% 
  dplyr::select(Acropora_023112) %>% 
  rownames_to_column() %>% 
  separate(rowname, into = c("Sample_ID", "Genotype", "Expression", "Treatment"), sep ="_") %>% 
  unite("Genotype", Genotype,Expression, sep = "_") %>% 
  mutate(Day = case_when(Treatment == "Initial" ~ 0,
                         Treatment == "Treated" ~ 28,
                         Treatment == "Untreated" ~ 28)) %>% 
  dplyr::select(!Treatment) %>% 
  full_join(., metadata) %>% 
  drop_na() %>% 
  # group_by(Day, Treatment, Genotype) %>% 
  # mutate(mean_expression = mean(Acropora_023112), se = standard_error(Acropora_023112)) %>% 
  ggplot(., aes(x=Day, y=Acropora_023112, color = Treatment, fill = Treatment)) +
  geom_point(alpha=0.2) +
  theme_classic() +
  geom_smooth(method = "lm", mapping = aes(color = Treatment), se=FALSE) +
  scale_color_manual(values = c( "#FF3333", "#00CCCC"))+
  labs(y = "Vst-Normalized Expression Level", title = "Acropora_023112 = Histone H2A") +
  theme(text = element_text(size = 13))+
  theme(legend.position="none") 
```

## nitric oxide synthase activity linear plot
```{r}
#Acropora_019471 = 	nitric-oxide synthase activity l2fc -2.052
plot2<-t(vsd) %>% 
  as.data.frame() %>% 
  dplyr::select(Acropora_019471) %>% 
  rownames_to_column() %>% 
  separate(rowname, into = c("Sample_ID", "Genotype", "Expression", "Treatment"), sep ="_") %>% 
  unite("Genotype", Genotype,Expression, sep = "_") %>% 
  mutate(Day = case_when(Treatment == "Initial" ~ 0,
                         Treatment == "Treated" ~ 28,
                         Treatment == "Untreated" ~ 28)) %>% 
  dplyr::select(!Treatment) %>% 
  full_join(., metadata) %>% 
  drop_na() %>% 
  ggplot(., aes(x=Day, y=Acropora_019471, color = Treatment, fill = Treatment)) +
  geom_point(alpha=0.2) +
  theme_classic() +
  geom_smooth(method = "lm", mapping = aes(color = Treatment), se=FALSE) +
  scale_color_manual(values = c( "#FF3333", "#00CCCC"))+
  labs(y = "Vst-Normalized Expression Level", title = "Acropora_019471 = Nitric-oxide synthase activity") +
  theme(text = element_text(size = 13)) +
  theme(legend.position="none") 
```

```{r}
plot_grid(plot1,plot2)
ggsave("specific_genes_linear.pdf", width = 12, height = 5)
```


# dds normalized gene counts
```{r}
dds <- estimateSizeFactors(dds)
normalized_counts <- counts(dds, normalized=TRUE)
```

## boxplot of Acropora_023112 = histone H2A 
```{r}
Acropora_023112_counts <- normalized_counts["Acropora_023112",]

df_Acropora_023112 <- data.frame(expression_level = Acropora_023112_counts)

df_Acropora_023112 %>% 
  rownames_to_column(var="Sample") -> df_Acropora_023112

design %>% 
  rownames_to_column(var="Sample") -> design
  
full_join(design, df_Acropora_023112) %>% 
  filter(!Treatment == "Initial") %>% 
  ggplot(., aes(x=Treatment, y=expression_level, fill=Treatment)) +
  geom_boxplot() +
  theme_classic() +
  scale_fill_manual(values = c( "#FF3333", "#00CCCC"))+
  labs(y = "Normalized Expression Level") +
  theme(text = element_text(size = 15))

```

## boxplot of Acropora_019471 = 	nitric-oxide synthase activity
```{r}
Acropora_019471_counts <- normalized_counts["Acropora_019471",]

df_Acropora_019471 <- data.frame(expression_level = Acropora_019471_counts)

df_Acropora_019471 %>% 
  rownames_to_column(var="Sample") -> df_Acropora_019471
 
full_join(design, df_Acropora_019471) %>% 
  filter(!Treatment == "Initial") %>% 
  ggplot(., aes(x=Treatment, y=expression_level, fill=Treatment)) +
  geom_boxplot() +
  theme_classic() +
  scale_fill_manual(values = c( "#FF3333", "#00CCCC"))+
  labs(y = "Normalized Expression Level") +
  theme(text = element_text(size = 15))

```


# ARCHIVE - gene expression / slowburn survival correlations

```{r}
design %>% 
  filter(!Treatment == "Initial") -> design_TreatedvUntreated # 20 samples 

Acer_slowburn_survival$Genotype <- gsub("-", "_", Acer_slowburn_survival$Colony)
```


## Acropora_023112 = histone H2A
```{r}
as.data.frame(normalized_counts) %>% 
  rownames_to_column(var="gene_ID") %>% 
  filter(gene_ID == "Acropora_023112") %>% 
  pivot_longer(Acer.005:Acer.150, names_to = "Sample") %>% 
  right_join(., design_TreatedvUntreated) %>% 
  select(!c("gene_ID", "Sample", "Sample.ID", "Date.Sampled")) -> normalized_counts_Acropora_023112

Acer_slowburn_survival %>% 
  right_join(., normalized_counts_Acropora_023112) %>% 
  group_by(Treatment) %>% 
  do(tidy(lm(Percent_Survival_Day5 ~ value, data = .))) # for p-values (neither correlation is significant)

Acer_slowburn_survival %>% 
  right_join(., normalized_counts_Acropora_023112) %>% 
  group_by(Treatment) %>%
  summarize(correlation = cor(Percent_Survival_Day5, value)) #for R2 values

p1<-Acer_slowburn_survival %>% 
  right_join(., normalized_counts_Acropora_023112) %>% 
  group_by(Treatment,Genotype) %>% 
  ggplot(., aes(x=Percent_Survival_Day5, y = value, color= Treatment)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, aes(group = Treatment)) + # Add linear model lines without confidence interval
  theme_classic() +
  scale_color_manual(values = c( "#FF3333", "#00CCCC"))+
  labs(title = "Acropora_023112 = histone H2A", x = "% Survival at 32ºC by Day 5", y = "Normalized Expression Level") +
  facet_wrap(~Treatment, scales="free_x")+
  theme(legend.position = "none")


Acer_slowburn_survival %>% 
  right_join(., normalized_counts_Acropora_023112) %>% 
  group_by(Treatment,Genotype) %>% 
  mutate(mean_expression = mean(value)) %>% 
  ggplot(., aes(x=Percent_Survival_Day5, y = mean_expression, color= Treatment)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, aes(group = Treatment)) + # Add linear model lines without confidence interval
  theme_classic() +
  scale_color_manual(values = c( "#FF3333", "#00CCCC"))+
  labs(y = "Mean Expression Level") +
  theme(text = element_text(size = 15))+
  facet_wrap(~Treatment, scales="free_x")
```

## Acropora_019471 = 	nitric-oxide synthase activity
```{r}
as.data.frame(normalized_counts) %>% 
  rownames_to_column(var="gene_ID") %>% 
  filter(gene_ID == "Acropora_019471") %>% 
  pivot_longer(Acer.005:Acer.150, names_to = "Sample") %>% 
  right_join(., design_TreatedvUntreated) %>% 
  select(!c("gene_ID", "Sample", "Sample.ID", "Date.Sampled")) -> normalized_counts_Acropora_019471

Acer_slowburn_survival %>% 
  right_join(., normalized_counts_Acropora_019471) %>% 
  group_by(Treatment) %>% 
  do(tidy(lm(Percent_Survival_Day5 ~ value, data = .))) # for p-values (neither correlation is significant)

Acer_slowburn_survival %>% 
  right_join(., normalized_counts_Acropora_019471) %>% 
  group_by(Treatment) %>%
  summarize(correlation = cor(Percent_Survival_Day5, value)) #for R2 values

p2<-Acer_slowburn_survival %>% 
  right_join(., normalized_counts_Acropora_019471) %>% 
  group_by(Treatment,Genotype) %>% 
  ggplot(., aes(x=Percent_Survival_Day5, y = value, color= Treatment)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, aes(group = Treatment)) + # Add linear model lines without confidence interval
  theme_classic() +
  scale_color_manual(values = c( "#FF3333", "#00CCCC"))+
  labs(title = "Acropora_019471 = nitric-oxide synthase activity", x = "% Survival at 32ºC by Day 5", y = "Normalized Expression Level") +
  facet_wrap(~Treatment, scales="free_x")+
  theme(legend.position = "none")

Acer_slowburn_survival %>% 
  right_join(., normalized_counts_Acropora_019471) %>% 
  group_by(Treatment,Genotype) %>% 
  mutate(mean_expression = mean(value)) %>% 
  ggplot(., aes(x=Percent_Survival_Day5, y = mean_expression, color= Treatment)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, aes(group = Treatment)) + # Add linear model lines without confidence interval
  theme_classic() +
  scale_color_manual(values = c( "#FF3333", "#00CCCC"))+
  labs(y = "Mean Expression Level") +
  theme(text = element_text(size = 15)) +
  facet_wrap(~Treatment, scales="free_x")
```

## Acropora_010783 (green fluorescent protein)
```{r}
as.data.frame(normalized_counts) %>% 
  rownames_to_column(var="gene_ID") %>% 
  filter(gene_ID == "Acropora_010783") %>% 
  pivot_longer(Acer.005:Acer.150, names_to = "Sample") %>% 
  right_join(., design_TreatedvUntreated) %>% 
  select(!c("gene_ID", "Sample", "Sample.ID", "Date.Sampled")) -> normalized_counts_Acropora_010783

Acer_slowburn_survival %>% 
  right_join(., normalized_counts_Acropora_010783) %>% 
  group_by(Treatment) %>% 
  do(tidy(lm(Percent_Survival_Day5 ~ value, data = .))) # for p-values (neither correlation is significant)

Acer_slowburn_survival %>% 
  right_join(., normalized_counts_Acropora_010783) %>% 
  group_by(Treatment) %>%
  summarize(correlation = cor(Percent_Survival_Day5, value)) #for R2 values

p3<-Acer_slowburn_survival %>% 
  right_join(., normalized_counts_Acropora_010783) %>% 
  group_by(Treatment,Genotype) %>% 
  ggplot(., aes(x=Percent_Survival_Day5, y = value, color= Treatment)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, aes(group = Treatment)) + # Add linear model lines without confidence interval
  theme_classic() +
  scale_color_manual(values = c( "#FF3333", "#00CCCC"))+
   labs(title = "Acropora_010783 = green fluorescent protein", x = "% Survival at 32ºC by Day 5", y = "Normalized Expression Level") +
  facet_wrap(~Treatment, scales="free_x")+
  theme(legend.position = "none")

Acer_slowburn_survival %>% 
  right_join(., normalized_counts_Acropora_010783) %>% 
  group_by(Treatment,Genotype) %>% 
  mutate(mean_expression = mean(value)) %>% 
  ggplot(., aes(x=Percent_Survival_Day5, y = mean_expression, color= Treatment)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, aes(group = Treatment)) + # Add linear model lines without confidence interval
  theme_classic() +
  scale_color_manual(values = c( "#FF3333", "#00CCCC"))+
  labs(y = "Mean Expression Level") +
  theme(text = element_text(size = 15)) +
  facet_wrap(~Treatment, scales="free_x")
```


## Acropora_000238 = nucleosome (Nucleosomes wrap and compact DNA into chromatin, limiting DNA accessibility to the cellular machineries which require DNA as a template)
```{r}
as.data.frame(normalized_counts) %>% 
  rownames_to_column(var="gene_ID") %>% 
  filter(gene_ID == "Acropora_000238") %>% 
  pivot_longer(Acer.005:Acer.150, names_to = "Sample") %>% 
  right_join(., design_TreatedvUntreated) %>% 
  select(!c("gene_ID", "Sample", "Sample.ID", "Date.Sampled")) -> normalized_counts_Acropora_000238

Acer_slowburn_survival %>% 
  right_join(., normalized_counts_Acropora_000238) %>% 
  group_by(Treatment) %>% 
  do(tidy(lm(Percent_Survival_Day5 ~ value, data = .))) # for p-values (neither correlation is significant)

Acer_slowburn_survival %>% 
  right_join(., normalized_counts_Acropora_000238) %>% 
  group_by(Treatment) %>%
  summarize(correlation = cor(Percent_Survival_Day5, value)) #for R2 values

p4<-Acer_slowburn_survival %>% 
  right_join(., normalized_counts_Acropora_000238) %>% 
  group_by(Treatment,Genotype) %>% 
  ggplot(., aes(x=Percent_Survival_Day5, y = value, color= Treatment)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, aes(group = Treatment)) + # Add linear model lines without confidence interval
  theme_classic() +
  scale_color_manual(values = c( "#FF3333", "#00CCCC"))+
  labs(title = "Acropora_000238 = nucleosome", x = "% Survival at 32ºC by Day 5", y = "Normalized Expression Level") +
  facet_wrap(~Treatment, scales="free_x")+
  theme(legend.position = "none")

Acer_slowburn_survival %>% 
  right_join(., normalized_counts_Acropora_000238) %>% 
  group_by(Treatment,Genotype) %>% 
  mutate(mean_expression = mean(value)) %>% 
  ggplot(., aes(x=Percent_Survival_Day5, y = mean_expression, color= Treatment)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, aes(group = Treatment)) + # Add linear model lines without confidence interval
  theme_classic() +
  scale_color_manual(values = c( "#FF3333", "#00CCCC"))+
  labs(y = "Mean Expression Level") +
  theme(text = element_text(size = 15)) +
  facet_wrap(~Treatment, scales="free_x")
```

## Acropora_015765 = transmembrane receptor protein tyrosine kinase activity
```{r}
as.data.frame(normalized_counts) %>% 
  rownames_to_column(var="gene_ID") %>% 
  filter(gene_ID == "Acropora_015765") %>% 
  pivot_longer(Acer.005:Acer.150, names_to = "Sample") %>% 
  right_join(., design_TreatedvUntreated) %>% 
  select(!c("gene_ID", "Sample", "Sample.ID", "Date.Sampled")) -> normalized_counts_Acropora_015765

Acer_slowburn_survival %>% 
  right_join(., normalized_counts_Acropora_015765) %>% 
  group_by(Treatment) %>% 
  do(tidy(lm(Percent_Survival_Day5 ~ value, data = .))) # for p-values (neither correlation is significant)

Acer_slowburn_survival %>% 
  right_join(., normalized_counts_Acropora_015765) %>% 
  group_by(Treatment) %>%
  summarize(correlation = cor(Percent_Survival_Day5, value)) #for R2 values

p5<-Acer_slowburn_survival %>% 
  right_join(., normalized_counts_Acropora_015765) %>% 
  group_by(Treatment,Genotype) %>% 
  ggplot(., aes(x=Percent_Survival_Day5, y = value, color= Treatment)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, aes(group = Treatment)) + # Add linear model lines without confidence interval
  theme_classic() +
  scale_color_manual(values = c( "#FF3333", "#00CCCC"))+
  labs(title = "Acropora_015765 = transmembrane receptor protein tyrosine kinase activity", x = "% Survival at 32ºC by Day 5", y = "Normalized Expression Level") +
  facet_wrap(~Treatment, scales="free_x")+
  theme(legend.position = "none")

Acer_slowburn_survival %>% 
  right_join(., normalized_counts_Acropora_015765) %>% 
  group_by(Treatment,Genotype) %>% 
  mutate(mean_expression = mean(value)) %>% 
  ggplot(., aes(x=Percent_Survival_Day5, y = mean_expression, color= Treatment)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, aes(group = Treatment)) + # Add linear model lines without confidence interval
  theme_classic() +
  scale_color_manual(values = c( "#FF3333", "#00CCCC"))+
  labs(y = "Mean Expression Level") +
  theme(text = element_text(size = 15)) +
  facet_wrap(~Treatment, scales="free_x")

```

## Acropora_017542 = 	Belongs to the small heat shock protein (HSP20) family
```{r}
as.data.frame(normalized_counts) %>% 
  rownames_to_column(var="gene_ID") %>% 
  filter(gene_ID == "Acropora_017542") %>% 
  pivot_longer(Acer.005:Acer.150, names_to = "Sample") %>% 
  right_join(., design_TreatedvUntreated) %>% 
  select(!c("gene_ID", "Sample", "Sample.ID", "Date.Sampled")) -> normalized_counts_Acropora_017542

Acer_slowburn_survival %>% 
  right_join(., normalized_counts_Acropora_017542) %>% 
  group_by(Treatment) %>% 
  do(tidy(lm(Percent_Survival_Day5 ~ value, data = .))) # for p-values (neither correlation is significant)

Acer_slowburn_survival %>% 
  right_join(., normalized_counts_Acropora_017542) %>% 
  group_by(Treatment) %>%
  summarize(correlation = cor(Percent_Survival_Day5, value)) #for R2 values

p6<-Acer_slowburn_survival %>% 
  right_join(., normalized_counts_Acropora_017542) %>% 
  group_by(Treatment,Genotype) %>% 
  ggplot(., aes(x=Percent_Survival_Day5, y = value, color= Treatment)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, aes(group = Treatment)) + # Add linear model lines without confidence interval
  theme_classic() +
  scale_color_manual(values = c( "#FF3333", "#00CCCC"))+
  labs(title = "Acropora_017542 = Belongs to the small heat shock protein (HSP20) family", x = "% Survival at 32ºC by Day 5", y = "Normalized Expression Level") +
  facet_wrap(~Treatment, scales="free_x") +
  theme(legend.position = "none")


Acer_slowburn_survival %>% 
  right_join(., normalized_counts_Acropora_017542) %>% 
  group_by(Treatment,Genotype) %>% 
  mutate(mean_expression = mean(value)) %>% 
  ggplot(., aes(x=Percent_Survival_Day5, y = mean_expression, color= Treatment)) +
  geom_point() +
  geom_smooth(method = "lm", se = FALSE, aes(group = Treatment)) + # Add linear model lines without confidence interval
  theme_classic() +
  scale_color_manual(values = c( "#FF3333", "#00CCCC"))+
  labs(y = "Mean Expression Level") +
  theme(text = element_text(size = 15)) +
  facet_wrap(~Treatment, scales="free_x")
```


## plot all correlation plots together
```{r}
plot_grid(p1, p2, p3, p4, p5, p6, ncol = 3)
ggsave("specificgenes_correlationsurvivorship.pdf", width = 10, height = 6)
```

